File: day-03-2/ruby/answer.rb
# frozen_string_literal: true

class Answer
  def self.read_input_file
    File.read('input.txt')
  end

  def count_ones(data, index)
    data.reduce(0) { |count, number| count + number[index].to_i }
  end

  def most_common(data, index)
    count_ones(data, index) >= data.size / 2.0 ? '1' : '0'
  end

  def least_common(data, index)
    count_ones(data, index) >= data.size / 2.0 ? '0' : '1'
  end

  def filter_candidates(data, function)
    index = 0

    while data.size > 1
      target = send(function, data, index)
      data.select! { |number| number[index] == target }
      index += 1
    end
  end

  def main
    diagnostic_report = self.class.read_input_file.split("\n")

    oxygen_candidates = diagnostic_report.dup
    co2_candidates = diagnostic_report.dup

    filter_candidates(oxygen_candidates, 'most_common')
    filter_candidates(co2_candidates, 'least_common')

    oxygen_generator_rating = oxygen_candidates.first.to_i(2)
    co2_scrubber_rating = co2_candidates.first.to_i(2)

    life_support_rating = oxygen_generator_rating * co2_scrubber_rating
    life_support_rating
  end
end

puts Answer.new.main

File: day-03-2/java/src/main/java/com/adventofcode/Answer.java
package com.adventofcode;

import java.io.FileInputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

class Answer {

    public static String readInputFile() {
        StringBuilder strBuilder = new StringBuilder();
        int readVar;

        try (FileInputStream fileInput = new FileInputStream("input.txt")) {
            do {
                readVar = fileInput.read();
                if (readVar != -1) {
                    strBuilder.append((char) readVar);
                }
            } while (readVar != -1);
        } catch (IOException e) {
            System.out.println("I/O Error: " + e);
            return "";
        }

        return strBuilder.toString();
    }

    public List<String> filterByCharAndIndex(List<String> list, char chr, int index) {
        return list.stream()
                   .filter(binNumber -> binNumber.charAt(index) == chr)
                   .collect(Collectors.toList());
    }

    public int indexOneCount(List<String> list, int index) {
        return (int) list.stream()
                         .map(binNumber -> binNumber.charAt(index))
                         .filter(bit -> bit == '1')
                         .count();
    }

    public int solution() {
        String[] diagnosticReport = readInputFile().split("\n");
        List<String> oxygenCandidates = new ArrayList<>(Arrays.asList(diagnosticReport));
        List<String> co2Candidates = new ArrayList<>(Arrays.asList(diagnosticReport));
        int oxygenGeneratorRating;
        int co2ScrubberRating;
        int lifeSupportRating;
        int index;

        index = 0;
        while (oxygenCandidates.size() > 1) {
            int indexOneCount = indexOneCount(oxygenCandidates, index);
            char mostCommon = indexOneCount >= oxygenCandidates.size() / 2.0 ? '1' : '0';
            oxygenCandidates = filterByCharAndIndex(oxygenCandidates, mostCommon, index++);
        }

        index = 0;
        while (co2Candidates.size() > 1) {
            int indexOneCount = indexOneCount(co2Candidates, index);
            char leastCommon = indexOneCount >= co2Candidates.size() / 2.0 ? '0' : '1';
            co2Candidates = filterByCharAndIndex(co2Candidates, leastCommon, index++);
        }

        oxygenGeneratorRating = Integer.parseInt(oxygenCandidates.get(0), 2);
        co2ScrubberRating = Integer.parseInt(co2Candidates.get(0), 2);

        lifeSupportRating = oxygenGeneratorRating * co2ScrubberRating;
        return lifeSupportRating;
    }

    public static void main(String[] args) {
        System.out.println(new Answer().solution());
    }

}
